src/ibek/ioc_factory.py

```diff
@@ -0,0 +1,225 @@
+
+        # extract the ioc instance yaml into a dict
+        ioc_instance_dict = YAML(typ="safe").load(ioc_instance_yaml)
+
+        if ioc_instance_dict is None or "ioc_name" not in ioc_instance_dict:
+            raise RuntimeError(
+                f"Failed to load a valid ioc config from {ioc_instance_yaml}"
+            )
+
+        # extract the ioc name into UTILS for use in jinja renders
+        name = UTILS.render({}, ioc_instance_dict["ioc_name"])
+        UTILS.set_ioc_name(name)
+        ioc_instance_dict["ioc_name"] = name
+
+        # Create an IOC instance from the instance dict and the model
+        ioc_instance = ioc_model(**ioc_instance_dict)
+        # post processing to insert SubEntity instances
+        self._process_collections(ioc_instance)
+
+        return ioc_instance
+
+    def ioc_create_model(self, definitions: List[Path]) -> Type[IOC]:
+        """
+        Take a list of definitions YAML and create an IOC model from it
+        """
+        entity_models = []
+
+        clear_entity_model_ids()
+        for definition in definitions:
+            support_dict = YAML(typ="safe").load(definition)
+
+            Support.model_validate(support_dict)
+
+            # deserialize the support module definition file
+            support = Support(**support_dict)
+            # make Entity classes described in the support module definition file
+            entity_models += self._make_entity_models(support)
+
+        # Save the schema for IOC
+        model = self._make_ioc_model(entity_models)
+
+        return model
+
+    def _process_collections(self, ioc_instance: IOC):
+        """
+        Process all the sub entity collections in the IOC instance
+        """
+        all_entities: List[Entity] = []
+
+        def scan_sub_entities(entity: Entity):
+            # recursive function to scan for SubEntities in an entity
+            definition = entity.__definition__
+
+            # add the parent standard entity
+            all_entities.append(entity)
+
+            # add in SubEntities if any
+            for sub_entity in definition.sub_entities:
+                # find the Entity Class that the SubEntity represents
+                entity_cls = self._entity_classes[sub_entity.type]
+                # get the SubEntity arguments
+                sub_args_dict = sub_entity.model_dump()
+                # jinja render any references to parent Args in the SubEntity Args
+                for key, arg in sub_args_dict.items():
+                    sub_args_dict[key] = UTILS.render(entity, arg)
+                # cast the SubEntity to its concrete Entity subclass
+                cast_entity = entity_cls(**sub_args_dict)
+                # recursively scan the SubEntity for more SubEntities
+                scan_sub_entities(cast_entity)
+
+        for entity in ioc_instance.entities:
+            scan_sub_entities(entity)
+
+        ioc_instance.entities = all_entities
+
+    def _make_entity_model(
```

Comment:
As discussed IRL, it might be nice to split the `Entity` creation out of `IocFactory` and then pass in a fully populated entity models dict

Reply:
I'm on it.

Reply:
OK this is done. There is now an entity_factory as well as ioc_factory. They know not of each other but they both understand Lists of Entity Classes.

This requires a little more work to orchestrate the two as per this block of code which I feel looks OK:
https://github.com/epics-containers/ibek/blob/d11b619d243b8f3c615bd64cce0afe65a06390e7/src/ibek/ioc_cmds/commands.py#L81-L85

ALSO - entity_factory now is where the list of all classes is kept and it is therefore the place that process_collections resides because that needs the list of classes in order to cast the SubEntities. This seems like a good fit to me.

Finally all those pesky clear_entity_model_ids() have been reduced to one. In the constructor of EntityFactory. If you are starting a new set of entity classes then the old list of instances can be chucked and this does make all the tests work.

Reply:
@GDYendell do you think the above looks clear enough to understand when we come back to it in 5 years?

Reply:
thanks!